whitenoise/asgi.py

```diff
@@ -0,0 +1,86 @@
+from __future__ import annotations
+
+
+class AsgiWhiteNoise:
+    # This is the same block size as wsgiref.FileWrapper
+    BLOCK_SIZE = 8192
+
+    def __init__(self, whitenoise, application):
+        self.whitenoise = whitenoise
+        self.application = application
+
+    async def __call__(self, scope, receive, send):
+        static_file = None
+        if scope["type"] == "http":
+            if self.whitenoise.autorefresh:
+                static_file = self.whitenoise.find_file(scope["path"])
+            else:
+                static_file = self.whitenoise.files.get(scope["path"])
+        if static_file is None:
+            await self.application(scope, receive, send)
+        else:
+            await receive_request(receive)
+            request_headers = convert_asgi_headers(scope["headers"])
+            await serve_static_file(
+                send, static_file, scope["method"], request_headers, self.BLOCK_SIZE
+            )
+
+
+async def serve_static_file(send, static_file, method, request_headers, block_size):
+    response = static_file.get_response(method, request_headers)
+    try:
+        await send(
+            {
+                "type": "http.response.start",
+                "status": response.status.value,
+                "headers": convert_wsgi_headers(response.headers),
+            }
+        )
+        if response.file:
+            # We need to only read content-length bytes instead of the whole file,
+            # the difference is important when serving range requests.
+            content_length = int(dict(response.headers)["Content-Length"])
+            for block in read_file(response.file, content_length, block_size):
+                await send(
+                    {"type": "http.response.body", "body": block, "more_body": True}
+                )
```

Comment:
So ASGI has a "zero copy send" extension: https://asgi.readthedocs.io/en/latest/extensions.html#zero-copy-send

This is much more efficient than chunking and recombining the file in Python

That said, it looks like support isn't there yet:

* daphne - none
* hypercorn - none
* uvicorn - pending https://github.com/encode/uvicorn/pull/1210

So it might be worth creating a follow-up issue for this

Reply:
hypercorn supports it now https://gitlab.com/synodriver/hypercorn/-/tree/zerocopy

Reply:
Has anything changed in the "zero copy send" landscape?

Reply:
no

Reply:
> “零拷贝发送”领域有什么变化吗？

I've created a fork of hypercorn to add all extension support [here](https://github.com/nonebot/nonecorn), you can have a try, currently it's under development and should work with all asgi extension listed in asgi spec.

Reply:
Even when zero copy send gets merged into most ASGI webservers, there's still the Windows compatibility questions

Ref: https://discuss.python.org/t/support-for-os-sendfile-for-windows/25020/11